atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/ScalingSweepTaskSchedulerTest.java

```diff
@@ -0,0 +1,310 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.sweep.queue;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.atMost;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import static com.palantir.atlasdb.sweep.queue.ScalingSweepTaskScheduler.INITIAL_DELAY;
+import static com.palantir.logsafe.testing.Assertions.assertThat;
+
+import java.time.Duration;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.jmock.lib.concurrent.DeterministicScheduler;
+import org.junit.Before;
+import org.junit.Test;
+
+import com.google.common.util.concurrent.Uninterruptibles;
+import com.palantir.common.concurrent.PTExecutors;
+
+public class ScalingSweepTaskSchedulerTest {
+    private static final SweepIterationResult SUCCESS_HUGE = SweepIterationResults
+            .success(SweepQueueUtils.SWEEP_BATCH_SIZE);
+    private static final SweepIterationResult SUCCESS_LARGE = SweepIterationResults
+            .success(ScalingSweepTaskScheduler.BATCH_CELLS_HIGH_THRESHOLD);
+    private static final SweepIterationResult SUCCESS_MEDIUM = SweepIterationResults
+            .success(ScalingSweepTaskScheduler.BATCH_CELLS_LOW_THRESHOLD + 1);
+    private static final SweepIterationResult SUCCESS_SMALL = SweepIterationResults
+            .success(ScalingSweepTaskScheduler.BATCH_CELLS_LOW_THRESHOLD);
+    private static final SweepIterationResult SUCCESS_TINY = SweepIterationResults
+            .success(SweepDelay.BATCH_CELLS_LOW_THRESHOLD);
+    private static final long DELAY = 1L;
+    private static final long INITIAL_PAUSE = 5L;
+
+    private final DeterministicScheduler deterministicScheduler = new DeterministicScheduler();
+    private final SweepDelay delay = mock(SweepDelay.class);
+    private final Callable<SweepIterationResult> sweepIteration = mock(Callable.class);
+    private final AtomicBoolean schedulerEnabled = new AtomicBoolean(true);
+    private final ScalingSweepTaskScheduler scheduler = createScheduler(Duration.ZERO);
+
+    @Before
+    public void setup() {
+        when(delay.getInitialPause()).thenReturn(INITIAL_PAUSE);
+        when(delay.getNextPause(any(SweepIterationResult.class))).thenReturn(DELAY);
+    }
+
+    @Test
+    public void whenExpectedNumberOfEntriesIsSweptKeepReschedulingAfterDelay() throws Exception {
+        when(sweepIteration.call()).thenReturn(SUCCESS_MEDIUM);
+        scheduler.start(2);
+
+        deterministicScheduler.tick(INITIAL_DELAY + 8 * DELAY, TimeUnit.MILLISECONDS);
```

Comment:
nit: probably just factor out `runSweepIterations(iterations)` for ticking `INITIAL_DELAY + (iterations - 1) * DELAY`